lib/events.js

```diff
@@ -0,0 +1,184 @@
+  },
+
+  getRequestOptions: function(payload){
+    var options = {
+      // events must be sent with POST
+      method: "POST", 
+      url: this.endpoint,
+      headers:{
+        'Content-Type': 'application/json'
+      },
+      qs: {
+        access_token: this.access_token
+      },
+      body:JSON.stringify([payload]) //events are arrays
+    }
+    return options
+  },
+
+  /**
+     * Get the event payload to send to the events service
+     * Most payload properties are shared across all events
+     * @param {String} event the name of the event to send to the events service. Valid options are 'search.start', 'search.select', 'search.feedback'.
+     * @param {Object} geocoder a mapbox-gl-geocoder instance 
+     * @returns {Object} an event payload 
+     */
+  getEventPayload: function(event, geocoder){
+    var proximity;
+    if (!geocoder.options.proximity) proximity = null;
+    else proximity = [geocoder.options.proximity.longitude, geocoder.options.proximity.latitude ];
+
+    var center = (geocoder._map) ? geocoder._map.getCenter() : {lat: null, lng: null};
+    var zoom = (geocoder._map) ? geocoder._map.getZoom() : null;
+    return {
+      event: event,
+      created: +new Date(),
+      sessionIdentifier: this.sessionID,
+      country: this.countries, 
+      userAgent: this.userAgent,
+      language: this.language,
+      bbox: this.bbox,
+      types: this.types, 
+      endpoint: 'mapbox.places',
+      // fuzzyMatch: search.fuzzy, //todo  --> add to plugin
+      proximity: proximity,
+      limit: geocoder.options.limit,
+      // routing: search.routing, //todo --> add to plugin
+      queryString: geocoder.inputString,
+      mapCenterLat: center.lat,
+      mapCenterLng: center.lng,
+      mapZoom: zoom,
+      keyboardLocale: this.locale 
+    }
+  },
+
+  /**
+     * Wraps the request function for easier testing
+     * Make an http request and invoke a callback
+     * @param {Object} opts options describing the http request to be made
+     * @param {Function} callback the callback to invoke when the http request is completed
+     */
+  request: function(opts, callback){
+    request(opts, function(err, res){
+      return callback(err, res);
+    })
+  },
+
+  /**
+     * Handle an error that occurred while making a request
+     * @param {Object} err an error instance to log
+     */
+  handleError: function(err, callback){
+    if (callback) return callback(err);
+  },
+
+  /**
+     * Generate a session ID to be returned with all of the searches made by this geocoder instance
+     * ID is random and cannot be tracked across sessions
+     */
+  generateSessionID: function(){
+    var sha = crypto.createHash('sha256');
+    sha.update(Math.random().toString());
+    return sha.digest('hex')
+  },
+
+  /**
+     * Get a user agent string to send with the request to the events service
+     */
+  getUserAgent: function(){
+    return 'mapbox-gl-geocoder.' + this.version + "." + navigator.userAgent;
```

Comment:
When I test this locally with `yarn run start` it reports "undefined" here. Not sure if that's intended or not?